source/extensions/access_loggers/fluentd/config.cc

```diff
@@ -0,0 +1,80 @@
+#include "source/extensions/access_loggers/fluentd/config.h"
+
+#include <memory>
+
+#include "envoy/registry/registry.h"
+#include "envoy/server/filter_config.h"
+
+#include "source/common/common/logger.h"
+#include "source/common/config/utility.h"
+#include "source/common/formatter/substitution_format_string.h"
+#include "source/common/formatter/substitution_formatter.h"
+#include "source/common/protobuf/protobuf.h"
+#include "source/extensions/access_loggers/fluentd/fluentd_access_log_impl.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace AccessLoggers {
+namespace Fluentd {
+
+// Singleton registration via macro defined in envoy/singleton/manager.h
+SINGLETON_MANAGER_REGISTRATION(fluentd_access_logger_cache);
+
+FluentdAccessLoggerCacheSharedPtr
+getAccessLoggerCacheSingleton(Server::Configuration::ServerFactoryContext& context) {
+  return context.singletonManager().getTyped<FluentdAccessLoggerCacheImpl>(
+      SINGLETON_MANAGER_REGISTERED_NAME(fluentd_access_logger_cache), [&context] {
+        return std::make_shared<FluentdAccessLoggerCacheImpl>(
+            context.clusterManager(), context.scope(), context.threadLocal());
+      });
+}
+
+AccessLog::InstanceSharedPtr
+FluentdAccessLogFactory::createAccessLogInstance(const Protobuf::Message& config,
+                                                 AccessLog::FilterPtr&& filter,
+                                                 Server::Configuration::FactoryContext& context) {
+  const auto& proto_config = MessageUtil::downcastAndValidate<
+      const envoy::extensions::access_loggers::fluentd::v3::FluentdAccessLogConfig&>(
+      config, context.messageValidationVisitor());
+
+  absl::Status status = context.serverFactoryContext().clusterManager().checkActiveStaticCluster(
+      proto_config.cluster());
+  if (!status.ok()) {
+    throw EnvoyException(fmt::format("cluster '{}' was not found", proto_config.cluster()));
+  }
+
+  // Supporting nested object serialization is more complex with MessagePack.
+  // Using an already existing JSON formatter, and later converting the JSON string to a msgpack
+  // payload.
+  // TODO(ohadvano): Improve the formatting operation by creating a dedicated formatter that
+  //                 will directly serialize the record to msgpack payload.
+  Formatter::FormatterPtr json_formatter =
+      Formatter::SubstitutionFormatStringUtils::createJsonFormatter(proto_config.record(), true,
+                                                                    false, false);
+  FluentdFormatterPtr fluentd_formatter =
+      std::make_unique<FluentdFormatterImpl>(std::move(json_formatter));
+
```

Comment:
could be cleaner to check here if instance exists in cache here in fsctory method instead of doing check in `FluentdAccessLogImpl` constructor

Reply:
The cache maintains instances of ``FluentdAccessLoggerImpl`` which are per-thread but the ``FluentdAccessLog`` is composing the action of formatting a message and then dispatch it to the per-thread logger instance